prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,274 @@
+	Started      time.Time
+	Duration     time.Duration
+	Result       string
+}
+
+type JobHistoryTemplate struct {
+	OlderLink  string
+	NewerLink  string
+	LatestLink string
+	Name       string
+	Builds     []BuildData
+}
+
+func readObject(obj *storage.ObjectHandle) ([]byte, error) {
+	rc, err := obj.NewReader(context.Background())
+	if err != nil {
+		return []byte{}, fmt.Errorf("failed to get reader for GCS object: %v", err)
+	}
+	return ioutil.ReadAll(rc)
+}
+
+func readLatestBuild(bkt *storage.BucketHandle, root string) (int, error) {
+	path := path.Join(root, latestBuildFile)
+	data, err := readObject(bkt.Object(path))
+	if err != nil {
+		return -1, fmt.Errorf("failed to read latest build number: %v", err)
+	}
+	n, err := strconv.Atoi(string(data))
+	if err != nil {
+		return -1, fmt.Errorf("failed to parse latest build number: %v", err)
+	}
+	return n, nil
+}
+
+// resolve sym links into the actual log directory for a particular test run
+func resolveSymLink(bkt *storage.BucketHandle, symLink string) (string, error) {
+	data, err := readObject(bkt.Object(symLink))
+	if err != nil {
+		return "", fmt.Errorf("failed to read %s: %v", symLink, err)
+	}
+	// strip gs://<bucket-name> from global address `u`
+	u := string(data)
+	prefixRe := regexp.MustCompile("gs://.*?/")
+	return prefixRe.ReplaceAllString(u, ""), nil
+}
+
+func getDir(bkt *storage.BucketHandle, root, id string) (string, error) {
+	bAttrs, err := bkt.Attrs(context.Background())
+	if err != nil {
+		return "", fmt.Errorf("failed to get bucket name: %v", err)
+	}
+	bktName := bAttrs.Name
+	if strings.HasPrefix(root, logsPrefix) {
+		return path.Join(bktName, root, id), nil
+	}
+	symLink := path.Join(root, id+".txt")
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return "", fmt.Errorf("failed to resolve sym link: %v", err)
+	}
+	return path.Join(bktName, dir), nil
+}
+
+func getObject(bkt *storage.BucketHandle, root, id, fname string) *storage.ObjectHandle {
+	if strings.HasPrefix(root, logsPrefix) {
+		p := path.Join(root, id, fname)
+		return bkt.Object(p)
+	}
+	symLink := path.Join(root, id+".txt")
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return &storage.ObjectHandle{}
+	}
+	return bkt.Object(path.Join(dir, fname))
+}
+
+func fileExists(bkt *storage.BucketHandle, root, id, fname string) bool {
+	obj := getObject(bkt, root, id, fname)
+	_, err := obj.Attrs(context.Background())
+	return err == nil
+}
+
+func readStarted(bkt *storage.BucketHandle, root, id string) (jobs.Started, error) {
+	s := jobs.Started{}
+	sobj := getObject(bkt, root, id, "started.json")
+	sdata, err := readObject(sobj)
+	if err != nil {
+		return s, fmt.Errorf("failed to read started.json for build %s: %v", id, err)
+	}
+	err = json.Unmarshal(sdata, &s)
+	if err != nil {
+		return s, fmt.Errorf("failed to parse started.json for build %s: %v", id, err)
+	}
+	return s, nil
+}
+
+func readFinished(bkt *storage.BucketHandle, root, id string) (jobs.Finished, error) {
```

Comment:
This function could be combined with `readStarted` if it took 2 additional parameters
- `fileName string`
- `data interface{}`

and removed the first return value.

Reply:
poke @ibzib